### PR TITLE
feat: Consistent and sensible layout_config

### PR DESCRIPTION
**Pull request by tjdevries**
Moving to this PR from https://github.com/nvim-telescope/telescope.nvim/pull/823 because of some problems with being able to merge back, but I didn't want to lose the history from that PR.

This should also hopefully make it easier for people to comment on the PR and give us some feedback before merging.

cc @l-kershaw @Conni2461

### CODE REVIEW

File: lua/telescope/previewers/buffer_previewer.lua
Comment (Conni2461) [inline lua/telescope/previewers/buffer_previewer.lua:548]: This whole block needs to go. Probably a automerge issue. It reimplements `git_commit_diff` which is no longer a thing. (And it actually uses a stash call :rofl:)

File: lua/telescope/deprecated.lua
Comment (Conni2461) [inline lua/telescope/deprecated.lua:2]: deprecated is currently only checked for options directly passed to a builtin (and then picker) but we could also check if the setup table has one of the deprecated keys. Help people clean that up ;)

File: lua/telescope/config/resolve.lua
Comment (Conni2461) [inline lua/telescope/config/resolve.lua:163]: This resolves in "stupid" `<br>` in the docgen output. I never said my formatter is prefect :rofl: but these two lines will wrap anyways and the issue is fixed if you put the `{padding = `foo`} <br>` immediatly after the `:`

File: lua/telescope/config/resolve.lua
Comment (Conni2461) [inline lua/telescope/config/resolve.lua:191]: Same as above :)

File: doc/telescope.txt
Comment (Conni2461) [inline doc/telescope.txt:1245]: Here is a lot more documentation in the source file. Not really important for the enduser so we could either make the two lines into only 2 `-` or we make the other 19 lines into 3 `-` (with some `<br>`s) up to you

File: lua/telescope/config.lua
Comment (Conni2461) [inline lua/telescope/config.lua:47]: do we need this? I don't see the function used outside of `config` but if we do at some point we should put it in utils instead

File: lua/telescope/config.lua
Comment (Conni2461) [inline lua/telescope/config.lua:40]: Maybe we should just have all requires at the top

File: lua/telescope/config.lua
Comment (Conni2461) [inline lua/telescope/config.lua:187]: Some of the descriptions are not indented (I am not the biggest fan of starting at col 1 here) :smile:

File: lua/telescope/config.lua
Comment (Conni2461) [inline lua/telescope/config.lua:218]: Longer than 80

File: lua/telescope/config.lua
Comment (Conni2461) [inline lua/telescope/config.lua:243]: This whole part is weirdly indented. I was also thinking maybe we should make this more complete.
Like in the builtin.init i put in the pickers configuration where i showed off more mappings:
```lua
      mappings = {
        i = {
          ["<c-d>"] = require("telescope.actions").delete_buffer,
          -- or right hand side can also be a the name of the action as string
          ["<c-d>"] = "delete_buffer",
        },
      }
```

Maybe we should also show the string option here and e.g.
```lua
["jj"] = { "<esc>", type = "command" },
["kk"] = { "<cmd>echo \"Hello, World!\"<cr>", type = "command" },
```

File: lua/telescope/config.lua
Comment (Conni2461) [inline lua/telescope/config.lua:252]: To long

File: lua/telescope/config.lua
Comment (Conni2461) [inline lua/telescope/config.lua:258]: Stupid question. Do we want to switch to lua fzy as default sorter? Its faster, matches better? I know this is unrelated, currently just thinking about it :rofl:

File: lua/telescope/config.lua
Comment (Conni2461) [inline lua/telescope/config.lua:168]: Are we doing anything with that. I am currently confused by that.

If this thing is nil we are not calling the local `get_border_size` in layout_strategies.lua function and if i am using a theme (dropdown) its true, so `get_border_size` will be called. I can set it to false, obv and that works as expected but maybe we should set it to true on default. And give it a description. Thoughts?!

File: lua/telescope/pickers/layout_strategies.lua
Comment (Conni2461) [inline lua/telescope/pickers/layout_strategies.lua:74]: Missing type.

Also we can have real emmy documentation for local functions

File: lua/telescope/pickers/layout_strategies.lua
Comment (Conni2461) [inline lua/telescope/pickers/layout_strategies.lua:81]: For stuff like that we should probably slowly move to `vim.F.if_nil` Just to make it even more robust

File: lua/telescope/pickers/layout_strategies.lua
Comment (Conni2461) [inline lua/telescope/pickers/layout_strategies.lua:120]: typo

File: lua/telescope/pickers/layout_strategies.lua
Comment (Conni2461) [inline lua/telescope/pickers/layout_strategies.lua:146]: I don't think this is good explained. Maybe "amount of lines to scroll"

File: lua/telescope/pickers/layout_strategies.lua
Comment (Conni2461) [inline lua/telescope/pickers/layout_strategies.lua:562]: `bottom_pane` exists twice. They look identical on first glance. But one should still be removed :)

**Review by Conni2461 — CHANGES_REQUESTED** [review]
I am not really proud of that review. I nitpicked way to much about comments and documentation. And had not really much to say about code, because its already really really good. Also i need way more time to get into how the whole math works to make any valid comments about that.

The two main things that needs to be fixed is remove duplicated code:
- remove a mysterious new duplicated previewer with new name :rofl: 
- remove one of the duplicated `bottom_pane` in layout_strategies

That are the two things that need to be fixed before merge. All other things can be address at a later time. :)

I'll run the PR for the next week and will look for issues

File: lua/telescope/config.lua
Comment (l-kershaw) [reply lua/telescope/config.lua:47]: I think this was a leftover from my initial attempts at setting up tests in `layout_strategies_spec.lua`.
It can be safely removed.

File: lua/telescope/pickers/layout_strategies.lua
Comment (l-kershaw) [reply lua/telescope/pickers/layout_strategies.lua:146]: I think "The number of lines to scroll through the previewer" is the clearest way to phrase it

File: doc/telescope.txt
Comment (l-kershaw) [reply doc/telescope.txt:1245]: I agree that this is not as important for the end user, so would be fine to get rid if we want.

**Comment by l-kershaw** [comment]
@Conni2461
I agree with most of your suggestions, even if they are "nitpicky".
It's better to nitpick before we implement breaking changes to minimise confusion when it goes live.

I've 👍 for stuff that I simply agree with and commented where I have more to say.

---

Since we've moved to a new PR, I no longer have write access, so can't implement any of your suggested changes.
So the changes will either have to wait for @tjdevries to do them (not sure exactly when he's going away, but probably busy getting ready to go if not already away) or for someone else with write access to this branch.

**Comment by tjdevries** [comment]
Oh dang, I forgot that you wouldn't be able to write to this. Sorry, if I get on an actual computer this week I'll see if I can figure out a way to do that 😂

**Comment by Conni2461** [comment]
@tjdevries i don't see a way to only add him to one branch/PR, so we can either add him as collaborator to the repository or make him a member of the org. Its up to you :) I can add him, if you don't find a PC ;)

**Comment by tjdevries** [comment]
Yeah, go ahead and add. 😁 Thanks.

**Comment by l-kershaw** [comment]
Thanks for adding me to the org :)

---

I've sorted some of the simpler issues that @Conni2461 mentioned.
Some of the other ones I'm not sure the best way to deal with and/or need more discussion.

File: lua/telescope/config.lua
Comment (l-kershaw) [reply lua/telescope/config.lua]: I think that this should probably be just a boolean.

I've changed the default to `true` and added some documentation.
I'm happy with the situation now, but will wait for someone else to "Resolve" once they're happy with it 🙂

**Comment by l-kershaw** [comment]
I've implemented some of the other changes requested now.

---

The only two remaining are:
- checking `deprecated` at setup
- switching to "lua fzy" as the default sorter

I will have a think about how best to check `deprecated`  in setup over the next couple of days.
I don't have any strong feelings on changing the default sorter.

File: lua/telescope/deprecated.lua
Comment (l-kershaw) [inline lua/telescope/deprecated.lua:50]: We currently use both `nvim_err_write` and `log.warn` for deprecation messages.
I think we might also use `log.info` elsewhere (despite our log level being at "warn") and also just straight `error`.

Do we want to standardise which method we are using?

**Comment by l-kershaw** [comment]
@tjdevries @Conni2461
Do you guys have any thoughts on standardising our deprecation warnings (and if so, to what)?

Also, is there anything else that we want to change in this PR before it is ready to merge?

**Comment by Conni2461** [comment]
Yeah we can talk about standardising for error messages and warning but this is out of scope for this PR. For the deprecated.lua file i prefer one way of doing it, i have no preferences in which way. For other deprecated messages we have, for example in `actions/init.lua` we currently use `error` but tj already removed them in one PR that is not finished (don't ask me for the number :rofl:)

I have not looked at it yet again but i didn't have big issues while using it. I 100% prefer to get this merged before neovim 0.5 gets released. It would be awful if new people check this out, have a config they feel comfortable with and we change everything within a couple of days.

I could take another look at the source code tomorrow morning :)

**Comment by tjdevries** [comment]
I plan on merging this today or tomorrow. :)

**Review by Conni2461 — APPROVED** [review]

**Comment by dhruvmanila** [comment]
So, I finally tried this and there seems to be one issue, I could be wrong.

<details>
<summary>Minimal config</summary>

```vim
set nocompatible hidden laststatus=2
map Q <nop>

" Download the plugin manager
if !filereadable('/tmp/plug.vim')
  silent !curl -fsSLo /tmp/plug.vim
    \ https://raw.githubusercontent.com/junegunn/vim-plug/master/plug.vim
endif

source /tmp/plug.vim

call plug#begin('/tmp/plugged')
" Plugins to test
Plug 'nvim-telescope/telescope.nvim', {'branch': 'layout_config'}

call plug#end()

autocmd VimEnter * PlugClean! | PlugUpdate --sync | close

" Load the plugins (packadd <plugin_name>)
packadd telescope.nvim

" For lua only plugins
lua << EOF

require("telescope").setup {
  defaults = {
    layout_strategy = "horizontal",
    layout_config = {
      horizontal = {
        -- preview_width = 0.55,
      },
    },
  },
  pickers = {
    help_tags = {
      layout_config = {
        preview_width = 0.65,
      },
    },
  }
}

EOF

```

</details>

If I have the `defaults.layout_config.horizontal.preview_width` set then the `pickers.help_tags.layout_config.preview_width` is not being picked up. Settings for individual pickers should override the default one, right?

Using the above config and opening `Telescope help_tags` will give you the appropriate `preview_width` configured for that picker, but when you uncomment the line in the `defaults` table, the default preview_width will be taken into calculations.

| | |
| --- | ----------- |
| Screenshot 2021-07-01 at 14 54 18  |  Screenshot 2021-07-01 at 14 53 59 |

**Comment by Conni2461** [comment]
@l-kershaw thanks for your work here :)

**Pull request merged**
